Configure prio-limit for ceph exporter container

arttor opened the pull request:
Closes #14710 

Allows to set `--prio-limit` and `--stats-period` options to Ceph exporter from CephCluster CRD.

parth-gr commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -406,6 +406,24 @@ type MonitoringSpec struct {
+
+	// Ceph exporter configuration
+	// +optional
+	Exporter *CephExporterSpec `json:"exporter,omitempty"`
```
please update the CR in the examples folder with an example(comment it)

arttor replied:
done

parth-gr commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -406,6 +406,24 @@ type MonitoringSpec struct {
+	// Time to wait before sending requests again to exporter server (seconds)
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=600
+	// +kubebuilder:default=5
```
I remember we have issues setting default I API,
cc @BlaineEXE

parth-gr replied:
Okay, if it works we are fine, 
Blaine had some suggestions adding default in API is not that good. https://github.com/rook/rook/pull/14435#discussion_r1710034114

BlaineEXE replied:
I haven't tested it in a while, but I have seen that when Rook detects a new CR added, it gets ADDed with the zero value, then quickly gets UPDATEd with the default. Sometimes this results in a double reconcile. I'm not sure if that's still the case or not. 

Even if the scenario I described still exists, the worst that happens is that we apply the zero value, then reconcile again with the default value. As long as the zero value is safe to use, it shouldn't be an issue.

I think it would be good to understand what the zero value means here though. Does zero mean the wait time is (A) actually zero seconds? Is it (B) infinity seconds (i.e. resend disabled)? Does it mean the exporter (C) uses its own internal default? (D) other?

If it's (C), then maybe Rook doesn't need to specify kubebuilder:default at all. If it's (A), there probably isn't an issue. If it's (B), we should make sure with the Ceph team that changing from 0(infinity) to 5 doesn't have problematic edge case behaviors. (B) could risk some exported thing(s) not getting resent at all if the exporter doesn't handle that edge case well.

BlaineEXE replied:
From what I can tell looking at Ceph's code, I think the zero value means zero seconds. 

https://github.com/ceph/ceph/blob/ccdd877abaf707fabc7e0b8d209a421a2f841be0/src/exporter/DaemonMetricCollector.cc#L32-L47

Additionally, if we set 0 specifically, I think it will take 0 to be the value, rather than taking Ceph's internal default of 5s. I'm not absolutely certain, but this is a safe assumption.

I don't see anything in Ceph's code that limits this to 600s. Is there a need/reason to set this maximum? https://github.com/ceph/ceph/blob/51a89906dff333becb40508f6b80029646801f8c/src/common/options/ceph-exporter.yaml.in#L37-L45

Since 0 is a valid setting for Ceph, I think it makes sense to pass the 0 value on from Rook to Ceph if Rook has zero in the CR. We could make this a pointer to an int `*int` instead, so that we can simply not apply the argument if the user leaves this unset, but that might be more work than is truly needed.

I think this (removing the max limit) will be fine. @travisn wdyt? 
```go
	// +kubebuilder:validation:Minimum=0
	// +kubebuilder:default=5
```

parth-gr commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -406,6 +406,24 @@ type MonitoringSpec struct {
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=10
+	// +kubebuilder:default=5
+	PerfCountersPrioLimit int `json:"perfCountersPrioLimit,omitempty"`
```
if it is positive can we use `unit`

arttor replied:
thanks, done

parth-gr commented:
Please add the commit description

sp98 commented:
@arttor Please run `make codegen` and push the changes again. Due to a different issue, ci is not catching this.

sp98 commented:
> @arttor Please run `make codegen` and push the changes again. Due to a different issue, ci is not catching this.

This would update the `/pkg/apis/ceph.rook.io/v1/zz_generated.deepcopy.go` file with the API changes.

arttor commented:
> Please also update the doc ceph-cluster-crd.md with the next settings, thanks!

done

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -406,6 +406,24 @@ type MonitoringSpec struct {
+}
+
+type CephExporterSpec struct {
+	// Only performance counters greater than or equal to this option are fetched
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=10
+	// +kubebuilder:default=5
+	PerfCountersPrioLimit uint `json:"perfCountersPrioLimit,omitempty"`
```
Why maximum 10 here? I'll be honest that I don't have a good grasp of what this value represents, but I do see at least one instance a unit tests sets prio-limit=12. That suggests to me that 12 is a valid value. Ceph's config yaml doesn't list a maximum, and internally Ceph processes the priority as int64.

https://github.com/ceph/ceph/blob/ccdd877abaf707fabc7e0b8d209a421a2f841be0/src/exporter/DaemonMetricCollector.cc#L119-L121

I think it might be in our best interest to use the same data type as Ceph here (int64) and not set the maximum, unless you have deeper knowledge about what this value means that I'm missing.

FWIW, Ceph also doesn't limit this to positive integers. Can priorities be negative numbers? I don't know, but if there is doubt, it is better to not apply min/max values unless we are certain.

BlaineEXE replied:
If `prio` here represents linux kernel priorities, then it seems to me like -20 to 139 (seems kind of arbitrary) are valid values (from blog). But also seems like I am missing something since lower numbers are higher prio in most cases. With my own open questions, I would leave the min/max unset if I were working on this.

https://blogs.oracle.com/linux/post/task-priority

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -406,6 +406,24 @@ type MonitoringSpec struct {
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=600
+	// +kubebuilder:default=5
+	StatsDelaySeconds uint `json:"statsDelaySeconds,omitempty"`
```
This parses into an int64 in Ceph with no min/max behavior applied as far as I can tell. If we aren't 100% certain about the min/max values. I think it would be best to use the same type as Ceph and not set min/max.

BlaineEXE replied:
In Ceph, this is documented as `--stats-period`, or sometimes `exporter_stats_period`. I agree that `StatsDelaySeconds` is a clearer name, but I think we should consider keeping Rook and Ceph config naming comparable to help users better match Ceph docs to Rook docs.

I would suggest `StatsPeriodSeconds` might be a good value (including units is always a good practice), and it would be helpful to document that the value applies the `exporter_stats_period` config which does return search hits in the upstream Ceph docs. IMO, a link to Ceph docs likely isn't critical for this setting.

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -406,6 +406,24 @@ type MonitoringSpec struct {
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=10
+	// +kubebuilder:default=5
+	PerfCountersPrioLimit uint `json:"perfCountersPrioLimit,omitempty"`
```
Ceph documentation does not mention `--prio-limit` or `exporter_prio_limit`. Given that `exporter_stats_period` is the way Ceph documents the other value, I think it would be helpful for users here to document that this value will be applied to `exporter_stats_period`.

I think adding `PerfCounters` as a prefix adds some helpful clarity around what `PrioLimit` means, so I'm happy to keep that for the help it adds.

BlaineEXE reviewed (changes requested):
This PR is very clean, so from a code standpoint, I have no requests. 

As with most things in Rook, naming and careful data type considerations are the minor but hard things we need to watch.

arttor commented:
> This PR is very clean, so from a code standpoint, I have no requests.
> 
> As with many things in Rook, naming and careful data type considerations are the minor but hard things.
> 
> [update] I'll summarize: let's use `int64` for both of these (same as ceph) and remove min/max specs unless @arttor has more in-depth knowledge about relevant values that I don't have.

Thanks for the review, i am fully agree that we schould try to be transparent in communicaiton with Ceph and it is better to have the same naming and validation for config params. Changed values to `int64` and removed limits. Also renamed Delay to Period.
